x-template.js

```diff
@@ -1,271 +1,986 @@
+    const valueMalformed = Unforgiving.#invalidTransition(string, stringIndex, value);
+    const errorMessagesKey = valueMalformed
+      ? Unforgiving.#valueMalformedToErrorMessagesKey.get(valueMalformed)
+      : Unforgiving.#valueToErrorMessagesKey.get(value);
+    const errorMessage = Unforgiving.#errorMessages.get(errorMessagesKey);
+    const substringMessage = `See substring \`${notParsed}\`.`;
+    const parsedThroughMessage = `Your HTML was parsed through: \`${parsed}\`.`;
+    const message = `[${errorMessagesKey}] ${errorMessage}\n${substringMessage}\n${parsedThroughMessage}`;
+    throw new Error(message);
+  }
+
+  static #validateRawString(rawString) {
+    Unforgiving.#rawUnicodeOrHexadecimal.lastIndex = 0;
+    if (Unforgiving.#rawUnicodeOrHexadecimal.test(rawString)) {
+      const errorMessagesKey = Unforgiving.#namedErrorsToErrorMessagesKey.get('javascript-unicode');
+      const errorMessage = Unforgiving.#errorMessages.get(errorMessagesKey);
+      const substringMessage = `See (raw) substring \`${rawString.slice(0, Unforgiving.#rawUnicodeOrHexadecimal.lastIndex)}\`.`;
+      const message = `[${errorMessagesKey}] ${errorMessage}\n${substringMessage}`;
+      throw new Error(message);
+    }
+  }
+
+  static #validateExit(fragment, element) {
+    if (element.value !== fragment) {
+      const tagName = element.value[Unforgiving.#localName];
+      const errorMessagesKey = Unforgiving.#namedErrorsToErrorMessagesKey.get('missing-closing-tag');
+      const errorMessage = Unforgiving.#errorMessages.get(errorMessagesKey);
+      const substringMessage = `Missing a closing </${tagName}>.`;
+      throw new Error(`[${errorMessagesKey}] ${errorMessage}\n${substringMessage}`);
+    }
+  }
+
+  // https://html.spec.whatwg.org/multipage/named-characters.html
+  static #replaceHtmlEntities(originalContent) {
+    let content = originalContent;
+    Unforgiving.#htmlEntityStart.lastIndex = 0;
+    while (Unforgiving.#htmlEntityStart.test(content)) {
+      const contentIndex = Unforgiving.#htmlEntityStart.lastIndex - 2;
+      Unforgiving.#entity.lastIndex = contentIndex;
+      if (!Unforgiving.#entity.test(content)) {
+        const errorMessagesKey = Unforgiving.#namedErrorsToErrorMessagesKey.get('malformed-html-entity');
+        const errorMessage = Unforgiving.#errorMessages.get(errorMessagesKey);
+        const substringMessage = `See substring \`${originalContent}\`.`;
+        throw new Error(`[${errorMessagesKey}] ${errorMessage}\n${substringMessage}`);
       }
-      htmlStrings[iii] = string;
+      const encoded = content.slice(contentIndex, Unforgiving.#entity.lastIndex);
+      Unforgiving.#htmlEntityContainer.innerHTML = encoded;
+      const decoded = Unforgiving.#htmlEntityContainer.content.textContent;
+      content = content.replace(encoded, decoded);
+      Unforgiving.#htmlEntityStart.lastIndex = contentIndex + decoded.length;
     }
-    // Again, there might be a quote we need to slice off here still.
```

Comment:
As mentioned — removing that `Forgiving` comment has made the git diff super wonky here. Let me know if you want that back in 👌